test_conformance/subgroups/test_workgroup_non_uniform_vote.cpp

```diff
@@ -0,0 +1,471 @@
+// lowest subgroup local id
+struct ELECT
+{
+    static void gen(cl_int *x, cl_int *t, cl_int *m, int ns, int nw, int ng)
+    {
+        int non_uniform_size = ng % nw;
+        log_info("  sub_group_elect...\n");
+        if (non_uniform_size)
+        {
+            log_info("  non uniform work group size mode ON\n");
+        }
+        // no work here needed.
+    }
+
+    static int chk(cl_int *x, cl_int *y, cl_int *mx, cl_int *my, cl_int *m,
+                   int ns, int nw, int ng)
+    {
+        int ii, i, j, k, n;
+        int nj = (nw + ns - 1) / ns;
+        cl_int tr, rr;
+        int non_uniform_size = ng % nw;
+        ng = ng / nw;
+        if (non_uniform_size) ng++;
+        int last_subgroup_size = 0;
+
+
+        for (k = 0; k < ng; ++k)
+        { // for each work_group
+            if (non_uniform_size && k == ng - 1)
+            {
+                set_last_worgroup_params(non_uniform_size, nj, ns, nw,
+                                         last_subgroup_size);
+            }
+            for (j = 0; j < nw; ++j)
+            { // inside the work_group
+                i = m[4 * j + 1] * ns + m[4 * j];
+                my[i] = y[j]; // read device outputs for work_group
+            }
+
+            for (j = 0; j < nj; ++j)
+            { // for each subgroup
+                ii = j * ns;
+                if (last_subgroup_size && j == nj - 1)
+                {
+                    n = last_subgroup_size;
+                }
+                else
+                {
+                    n = ii + ns > nw ? nw - ii : ns;
+                }
+                rr = 0;
+                for (i = 0; i < n; ++i)
+                { // for each work_item in subgroup
+                  // sum of output values should be 1
+
+                    my[ii + i] > 0 ? rr += 1 : rr += 0;
+                }
+                // expectation is that only one elected returned true
+                tr = n <= NR_OF_ACTIVE_WORK_ITEMS ? 1 : 2;
+                if (rr != tr)
+                {
+                    log_error(
+                        "ERROR: sub_group_elect() mismatch for sub group %d in "
+                        "work group %d. Expected: %d Obtained: %d  \n",
+                        j, k, tr, rr);
+                    return TEST_FAIL;
+                }
+            }
+
+            x += nw;
+            y += nw;
+            m += 4 * nw;
+        }
+        log_info("  sub_group_elect... passed\n");
+        return TEST_PASS;
+    }
+};
+
+static const char *elect_source =
+    "__kernel void test_elect(const __global Type *in, __global int4 *xy, "
+    "__global Type *out)\n"
+    "{\n"
+    "    int gid = get_global_id(0);\n"
+    "    XY(xy,gid);\n"
+    "    Type x = in[gid];\n"
+    "    if (xy[gid].x < NR_OF_ACTIVE_WORK_ITEMS) {\n"
+    // one in subgroup true others false.
+    "        out[gid] = sub_group_elect();\n"
+    "    } else {\n"
+    "        out[gid] = sub_group_elect();\n"
```

Comment:
If work items for which `xy[gid].x < NR_OF_ACTIVE_WORK_ITEMS` is true are all active, and work items for which `xy[gid].x < NR_OF_ACTIVE_WORK_ITEMS` is false are not active, then surely no threads will ever execute the `else` case here.

I presume that this is not the case and that `NR_OF_ACTIVE_WORK_ITEMS` is the number of work items for which the `if` condition should be true, and for the remaining items the `else` case is taken. If this is the case, the naming of `NR_OF_ACTIVE_WORK_ITEMS` is _really_ confusing and should be changed.

Reply:
Given how `xy[gid].x` is initialised, this always elects the work item with subgroup local ID 0 in every subgroup.

To ensure adequate coverage, work items that do not have a subgroup local ID of 0 should be elected for some subgroups as part of this testing.

Preferably, this would test that work items with each possible subgroup local ID are electable on implementations having a maximum subgroup size of 128. The test must still work properly on implementations with a smaller maximum subgroup size.

It is important that this at least test that work items with the maximum subgroup local ID are electable, i.e. that with a maximum subgroup size of 128, if no other work item is active, when the work item with local ID 127 calls `sub_group_elect()`, 1 is returned.

It is important that the test ensures that work items with a subgroup local ID between 64 and 95 are electable, for implementations where this is applicable.

It is important that the test ensures that work items with a subgroup local ID between 32 and 63 are electable, for implementations where this is applicable.

Reply:
Although the testing here is an improvement on what preceded it, there are still substantial gaps in coverage. I realise while most of what I said here should be true, the comment that local ID 0 is tested in every subgroup only applies to the `if` block, not the `else` block.

Until [LLVM bug 46199](https://bugs.llvm.org/show_bug.cgi?id=46199) is fixed it is unhelpful not to have a mode in which convergence is _not_ tested.

In both the `if` and `else` blocks, the active subgroups as would be reported by `sub_group_ballot(1)` are a contiguous range. It is important that this test use non-contiguous ranges of active subgroup local IDs.

Supposing an implementation was written to elect _any_ work item for which the work item with the preceding subgroup local ID was inactive or for which the preceding subgroup local ID would be -1. (E.g. if using a ballot to implement the election, and given `sub_group_ballot(1).x == 0xff00ff00`, suppose the implementation elects _two_ work item, one with subgroup local ID 8 and the other with subgroup local ID 24.) Any such implementation would be buggy but would still pass this test.

Supposing also an implementation was written to elect the work item with the least significant subgroup local ID out of the contiguous group of work items with the most significant local IDs in the subgroup. (E.g. if using a ballot to implement the election, and given `sub_group_ballot(1).x == 0xff00ff00`, suppose the implementation elects _one_ work item, with subgroup local ID 24.) Any such implementation would be buggy but would still pass this test.

Reply:
Then should I prepare test which run `n = subgroup_size` times kernel with parameter `elect_item` changing from `0` to `subgroup_size - 1` ?
```
if (xy[gid].x == elect_item)
sub_group_elect();
```

Reply:
That's not quite what I'm saying. It is important to test with a variety of sets of _active_ but unelected work items in the test.

The sorts of bit pattern used in memory testers might give you inspiration, here, e.g. 0x00ffff00, 0xff00ff00, 0xff0000ff, 0xff00ff00, 0x0f0f0f0f, 0x0f0ff0f0, 0xaaaa5555, 0x5555aaaa, 0x55aaaa55, some so on and so forth.

This doesn't have to be quite as elaborate at that, though, it just needs to cover relevant cases. Others may be able to better prescribe the testing needed. If you put forward a proposal I would be happy to review it, but I cannot see how there would be a textbook answer as to how to test this.

Reply:
Please verify this proposal: https://github.com/KhronosGroup/OpenCL-CTS/pull/1074/commits/65c43f3375942f0c923df9501912efc88081dfce